src/core/unit.c

```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
         return 0;
 }
 
+static int unit_is_recursive_template_dependency(Unit *u, const char *name) {
+        _cleanup_free_ char *instance = NULL;
+        const char *fragment_path;
+        _cleanup_set_free_ Set *names = NULL;
+        int r;
+
+        assert(u);
+        assert(name);
+
+        /* If a template unit has a direct dependency on itself that contains the unit name as part of the
+         * template instance, this will almost certainly lead to infinite recursion as systemd will keep
+         * instantiating new instances of the template unit. https://github.com/systemd/systemd/issues/17602
+         * shows a good example of how this can happen in practice. To guard against this, we check for
+         * templates that depend on themselves and have the instantiated unit name as part of the template
+         * instance of the dependency.
+         */
+
+        if (!unit_name_prefix_equal(u->id, name))
+                return 0;
+
+        if (u->type != unit_name_to_type(name))
+                return 0;
+
+        r = unit_file_find_fragment(u->manager->unit_id_map, u->manager->unit_name_map, name, &fragment_path, &names);
+        if (r < 0)
+                return r;
+
+        if (!u->fragment_path || !fragment_path)
```

Comment:
smells as if the first half of this check should be moved before the unit_file_find_fragment() invocaiton, no?